### ISSUE TITLE
Add a way to flush data from memstores

### ISSUE THREAD

**Issue by rkarthik007** [yugabyte/yugabyte-db#43]
It would be good to add an RPC/cqlsh command by which we can flush all the data in the memstores onto disk.

**Comment by OlegLoginov** [comment]
Implemented in the Community Edition as the command: 'yb-admin flush_table <keyspace> <table_name> [timeout_in_seconds]'

**Comment by kriteshranjan** [comment]
Closing this issue based on AI validation. cc: @m-iancu , @VamsidharGunturu

**Comment by kriteshranjan** [comment]
AI validation notes (`yb-issue-43-analysis.md`):

# Analysis: yugabyte/yugabyte-db#43 (no Jira cross-link)

Issue: "Add a way to flush data from memstores" — kind/enhancement, filed
2018-02-06, still OPEN. Body: "It would be good to add an RPC/cqlsh command
by which we can flush all the data in the memstores onto disk."

Validated: 2026-08-13. Checkout: master @ b89bc6233565f83487971bcbef4bc5b8532a42f3
(read-only). Kind: feature request → behavior-delta validation (current
capability vs requested capability), not a crash repro.

## Verdict
NOT-A-BUG — the requested capability already exists on master (and has since
commit 4902d06cd1, 2018-02-20, two weeks after the issue was filed). Manual
memstore flush is exposed per-table (`yb-admin flush_table` /
`flush_table_by_id`), per-tserver (`yb-ts-cli flush_all_tablets` /
`flush_tablet`), and for the master sys catalog (`yb-admin flush_sys_catalog`),
all backed by dedicated admin RPCs and demonstrated end-to-end on
2026.1.1.0-b91. The only unimplemented sliver is a cqlsh built-in statement,
which the 2018-03-14 contributor comment already scoped out as a "next step"
and which no longer matters operationally.

## Root cause (current-behavior evidence — where the capability lives)

Requested: an RPC/command to flush memstore data to disk. Present on master:

**CLI entry points**
- `yb-admin flush_table` — registered at `src/yb/tools/yb-admin_cli.cc:3105`,
  action handler `flush_table_action` at `src/yb/tools/yb-admin_cli.cc:913`
  (args: `<table> [<timeout_in_seconds>] (default 20) [ADD_INDEXES]`, line 911-912).
- `yb-admin flush_table_by_id` — registered at `src/yb/tools/yb-admin_cli.cc:3106`,
  handler at `src/yb/tools/yb-admin_cli.cc:931`.
- `yb-admin flush_sys_catalog [leader_only]` — registered at
  `src/yb/tools/yb-admin_cli.cc:3107`, handler at `src/yb/tools/yb-admin_cli.cc:948`.
- `yb-ts-cli flush_tablet <tablet_id>` / `flush_all_tablets` — op constants at
  `src/yb/tools/ts-cli.cc:127-128`, dispatch at `src/yb/tools/ts-cli.cc:1145`
  and `:1152`, usage text at `:940-941`.

**RPC chain (per-table path)**
1. `ClusterAdminClient::FlushTables` (`src/yb/tools/yb-admin_client.cc:2031`)
   → `YBClient::FlushTables` (`src/yb/client/client.cc:783` by id, `:796` by name)
   → `YBClient::Data::FlushOrCompactTables` (`src/yb/client/client-internal.cc:1330`/`:1346`),
   which invokes the master admin RPC via `MasterAdminProxy::FlushTablesAsync`
   (`src/yb/client/client-internal.cc:1316`).
2. Master service: `rpc FlushTables` declared at `src/yb/master/master_admin.proto:332`
   (+ `IsFlushTablesDone` at `:333`), routed to `FlushManager::FlushTables`
   (`src/yb/master/flush_manager.cc:31`; service wiring
   `src/yb/master/master_admin_service.cc:137-139`). The FlushManager fans out
   one `AsyncFlushTablets` task per tablet server
   (`src/yb/master/flush_manager.cc:100-105`), which sends the tserver admin RPC
   `FlushTabletsAsync` (`src/yb/master/async_flush_tablets_task.cc:125`).
3. Tserver service: `rpc FlushTablets` declared at
   `src/yb/tserver/tserver_admin.proto:484`, handled by
   `TabletServiceAdminImpl::FlushTablets` (`src/yb/tserver/tablet_service.cc:2119`);
   supports both explicit `tablet_ids` and `all_tablets=true`
   (`src/yb/tserver/tablet_service.cc:2149-2161`). The FLUSH operation runs
   `TabletsFlusher::Flush` → `tablet->Flush(FlushMode::kAsync, ...,
   rocksdb::FlushReason::kAdminFlush)` (`src/yb/tserver/tablet_service.cc:1968-1970`)
   then waits (`:1973-1975`).
4. Storage: `Tablet::Flush` (`src/yb/tablet/tablet.cc:2620`) flushes the intents
   RocksDB (`:2646`) and regular RocksDB (`:2651`) memtables to SST files.

**yb-ts-cli path (per-tserver, no master involved)**
`TsAdminClient::FlushOrCompactsTabletsImpl` (`src/yb/tools/ts-cli.cc:669`)
builds `FlushTabletsRequestPB` with `all_tablets=true` when no tablet id given
(`:679-684`) and calls the same tserver `FlushTablets` RPC directly (`:696`).

**Docs**
- `docs/content/stable/admin/yb-admin.md:716` — `flush_table` reference
  (usage at `:725`, examples at `:740` and `:753`/`:766` for by-id form).
- `docs/content/stable/admin/yb-ts-cli.md:179` — `flush_all_tablets`;
  `:191` — `flush_tablet`.

**History** (issue filed 2018-02-06; all of this landed after)
- `4902d06cd1` (2018-02-20) "ENG-2964: Allow flushing all tablets of a given
  table" — original `yb-admin flush_table` + master `FlushTables` RPC. Matches
  the contributor comment on the issue (2018-03-14: "Implemented in the
  Community Edition as the command: 'yb-admin flush_table ...'").
- `8de51f46f8` (2020-03-16) — `flush_table_by_id` (#3814).
- `2842ff9ef7` (2020-05-16) — `yb-ts-cli flush_all_tablets` / `flush_tablet`
  (#2785); release-noted in `docs/content/stable/releases/ybdb-releases/end-of-life/v2.1.md:628`.

**Not implemented**: a cqlsh built-in flush statement — no flush token exists
in the YCQL grammar (`src/yb/yql/cql/ql/parser/parser_gram.y`). The issue body
asked for "RPC/cqlsh command" (either), and the RPC + CLI variants fully cover
the operational need.

## Trigger matrix (capability delta: requested vs present on master)

| # | Requested sub-capability | Predicted | Observed |
|---|---|---|---|
| 1 | RPC to flush memstores to disk | PRESENT (`master_admin.proto:332`, `tserver_admin.proto:484`) | PRESENT — `flush_table` exercised the full chain; tserver logged `Processing FlushTablets from ...` (`tablet_service.cc` handler) |
| 2 | CLI per-table flush | PRESENT (`yb-admin_cli.cc:3105-3106`) | PRESENT — `flush_table` in help; ran successfully (exit 0, "Flushed [yugabyte.t1 ...] tables.") |
| 3 | Flush ALL memstores on a tserver | PRESENT (`ts-cli.cc:128`, `:1152`) | PRESENT — `flush_all_tablets` ran successfully ("Successfully flushed all tablets") |
| 4 | Per-tablet flush | PRESENT (`ts-cli.cc:127`, `:1145`) | PRESENT in usage output (`flush_tablet <tablet_id>`); not individually executed — same RPC path as #3 with `tablet_ids` instead of `all_tablets` (`ts-cli.cc:679-684`) |
| 5 | Master sys-catalog flush | PRESENT (`yb-admin_cli.cc:3107`) | PRESENT — `flush_sys_catalog` ran successfully ("Successful on host ...") |
| 6 | cqlsh built-in flush statement | ABSENT (no flush token in `parser_gram.y`) | ABSENT (source-level evidence; not demoed) |

All six cells match prediction.

## Reproduction (capability demo, not a crash repro)

Image: `docker.io/yugabytedb/yugabyte:2026.1.1.0-b91-aarch64` (Apple Silicon,
podman). Cluster via `yugabyted start --daemon=false`.

```sh
podman run -d --name yb43 docker.io/yugabytedb/yugabyte:2026.1.1.0-b91-aarch64 \
  bash -c '/home/yugabyte/bin/yugabyted start --daemon=false'
sleep 30

# Control: commands exist in help/usage output
podman exec yb43 bash -c '/home/yugabyte/bin/yb-admin --master_addresses=$(hostname):7100 help \
  | grep -E "flush_table|flush_sys_catalog"; /home/yugabyte/bin/yb-ts-cli 2>&1 \
  | grep -E "flush_tablet|flush_all_tablets"'
#  20. flush_table <table> [<timeout_in_seconds>] (default 20) [ADD_INDEXES] ...
#  21. flush_table_by_id <table_id> ...
#  22. flush_sys_catalog [leader_only]
#   flush_tablet <tablet_id>
#   flush_all_tablets

# Data: 1000 rows into a 1-tablet YSQL table -> memstore only (0 SSTs)
podman exec yb43 bash -c '/home/yugabyte/bin/ysqlsh -h $(hostname) \
  -c "CREATE TABLE t1 (id int PRIMARY KEY, v text) SPLIT INTO 1 TABLETS;" \
  -c "INSERT INTO t1 SELECT g, repeat(chr(97+g%26),100) FROM generate_series(1,1000) g;"'
# BEFORE: every tablet dir under /root/var/data/yb-data/tserver/data/rocksdb has 0 *.sst

# Flush per-table
podman exec yb43 bash -c '/home/yugabyte/bin/yb-admin --master_addresses=$(hostname):7100 \
  flush_table ysql.yugabyte t1 60'
# -> "Flushed [yugabyte.t1 [ysql_schema=public] [000034e1000030008000000000004000]] tables." exit=0
# AFTER: tablet dir now contains 000010.sst (66457 bytes)
# tserver INFO log: "I0814 03:46:36.880896 ... tablet_service.cc:2143] ... Processing FlushTablets from 10.88.0.3:41371"

# Flush all tablets on the tserver (after 500 more rows)
podman exec yb43 bash -c '/home/yugabyte/bin/yb-ts-cli --server_address=$(hostname):9100 flush_all_tablets'
# -> "Successfully flushed all tablets" exit=0; second SST (000012.sst) appeared

# Sys catalog
podman exec yb43 bash -c '/home/yugabyte/bin/yb-admin --master_addresses=$(hostname):7100 flush_sys_catalog'
# -> "flush_sys_catalog: Successful on host 10.88.0.3:7100" exit=0

podman rm -f yb43
```

Success signature: SST-file count for the tablet goes 0 → 1 across the flush;
tserver log line `Processing FlushTablets from` (handler
`src/yb/tserver/tablet_service.cc:2119`; log statement at `:2144` on master,
`:2143` in the 2026.1.1.0 build — benign line drift).

## Versions tested
- Source: master @ b89bc6233565f83487971bcbef4bc5b8532a42f3 (all citations).
- Runtime demo: `2026.1.1.0-b91-aarch64` (PostgreSQL 15.12-YB-2026.1.1.0-b0).
- Master affected: N/A — master HAS the capability; the issue predates it
  (filed 2018-02-06, capability landed 2018-02-20 in 4902d06cd1).

## Proposed fix direction
None needed. Recommend maintainers close #43 as completed: the request was
implemented in 4902d06cd1 (2018-02-20) and extended by 8de51f46f8
(flush_table_by_id) and 2842ff9ef7 (yb-ts-cli flush_tablet/flush_all_tablets),
plus flush_sys_catalog for the master. The one literal sub-request never built
— a cqlsh built-in — was explicitly deferred in the 2018-03-14 issue comment
and is obsolete: operators flush via yb-admin/yb-ts-cli (both documented), and
YCQL deliberately avoids admin statements in the query language. If
maintainers want that sliver tracked, a fresh narrowly-scoped issue would be
cleaner than keeping this 2018 catch-all open.

## Test strategy
Existing coverage (no new tests needed):
- `src/yb/tools/yb-admin-test.cc:1908` — `flush_table` via CLI
  (AdminCliTest); `:1668`/`:1675` — `flush_sys_catalog` (all peers +
  leader_only).
- `src/yb/tools/yb-ts-cli-test.cc:1194-1224` — `flush_tablet`,
  `flush_all_tablets`, compact variants.
- `src/yb/master/flush_manager-test.cc:68` (`VerifyFlush`) and `:91`
  (RPC-failure handling) — FlushManager end-to-end via `client_->FlushTables`.
- `src/yb/tools/yb-admin-multi-master-test.cc:178` — multi-master
  flush_sys_catalog.
All exercised via public entry points (CLI / YBClient API).

## Confidentiality check
Clean. All content derives from the public GitHub issue, the public source
tree, official public container images, and a local synthetic demo (table
`t1`, generated data). No customer identifiers, PII, internal hostnames,
secrets, or internal memory-note references. Container IP `10.88.0.3` is
podman's default local bridge inside a throwaway container, not a real
deployment address.
